### PR TITLE
[ROCM] Add Thrust support

### PR DESCRIPTION
**Pull request by masahi**
This adds rocThrust support for our rocm backend, which can be used for sorting and scan.
https://github.com/ROCmSoftwarePlatform/rocThrust

Use "rocm -libs=thrust" to enable thrust based sort and scan.

`contrib/rocthrust/thrust.cc` code is identical to `contrib/thrust/thrust.cu` used by CUDA backend. The same thrust code works on both CUDA and AMD. 

@mbrookhart @csullivan @antinucleon

### CODE REVIEW

File: src/runtime/contrib/rocthrust/thrust.cc [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+          thrust_stable_sort_by_key<int, int>(keys_in, values_in, keys_out, values_out,
+                                              for_scatter);
+        } else if (value_dtype == "int64") {
+          thrust_stable_sort_by_key<int, int64_t>(keys_in, values_in, keys_out, values_out,
+                                                  for_scatter);
+        } else if (value_dtype == "float32") {
+          thrust_stable_sort_by_key<int, float>(keys_in, values_in, keys_out, values_out,
+                                                for_scatter);
+        } else {
+          LOG(FATAL) << "Unsupported value dtype: " << value_dtype;
```
Comment (tkonolige) [inline src/runtime/contrib/rocthrust/thrust.cc:224]: For all these errors, could you list the supported types?

**Comment by masahi** [comment]
@tkonolige @mbrookhart Thanks for review, it should be ready to go. This can be used for benchmarking the new TIR sort @mbrookhart is cooking.

**Review by mbrookhart — COMMENTED** [review]
Devil's advocate:

If I reformat the thrust.cu file with clang-format-10, like you said, they are identical except for the datatype errors you added and `#include <thrust/sequence.h>` in the amd version. 

Is there any reason to keep two copies around?

(otherwise LGTM, thank you!)

File: python/tvm/relay/op/strategy/rocm.py [hunk 2]
```diff
@@ -219,3 +221,93 @@ def batch_matmul_strategy_rocm(attrs, inputs, out_type, target):
...
         )
     return strategy
+
+
+def can_use_thrust(target, func_name):
+    return (
+        target.kind.name == "rocm"
+        and "thrust" in target.libs
+        and get_global_func(func_name, allow_missing=True)
+    )
```
Comment (csullivan) [inline python/tvm/relay/op/strategy/rocm.py:231]: Thoughts on perhaps moving this helper to `python/tvm/contrib/rocm.py`? There are a few examples of helpers in python/tvm/contrib that help with op strategy dispatch based on conditions like these.

Comment (csullivan) [reply python/tvm/relay/op/strategy/rocm.py:231]: Or given a few of your TODOs in this PR, like
```
# TODO(masahi): Check -libs=thrust option
```
maybe we want a common helper for use with both cuda and rocm?

**Review by csullivan — COMMENTED** [review]
Thanks @masahi for adding rocthrust support!

**Comment by masahi** [comment]
@mbrookhart Yes, I wanted to reuse `thrust.cu` as is, by adding the following to `ROCM.cmake`:

```
file(GLOB CONTRIB_THRUST_SRC src/runtime/contrib/thrust/*.cu)
```

But unlike for CUDA, cmake doesn't recognize that this .cu file needs to be compiled by the same CXX compiler (`hipcc`), so it will not be compiled. I think CMake has a special support for NVCC and .cu file.

File: python/tvm/relay/op/strategy/rocm.py (hunk 2 above)
Comment (masahi) [reply python/tvm/relay/op/strategy/rocm.py:231]: Yes, currently thrust dispatch is done in a ad hoc way, unlike cudnn/cublas etc where `-libs` option is required in addition to `USE_CUDNN=ON` etc. 

@tkonolige and I discussed this issue a bit and we agreed that we should enforce `-libs=thrust` to enable thrust offload too. For rocm, this requirement is already enforced in this PR, but for cuda it is not. That's what `TODO(masahi)` will going to fix in the next PR.

We can discuss what the common helper should be / where it should be in the next PR, would you agree?

Comment (csullivan) [reply python/tvm/relay/op/strategy/rocm.py:231]: Works for me. :shipit:

**Comment by masahi** [comment]
Maybe pytorch people have a solution to the duplication problem. But my cmake-fu is too low and I have no interest in learning about cmake :)

**Comment by mbrookhart** [comment]
@tkonolige @csullivan Any other change requests or are you happy?

**Comment by tkonolige** [comment]
@masahi You can use `set_source_files_properties(src/runtime/contrib/thrust/thrust.cu PROPERTIES LANGUAGE CXX)`.

**Review by tkonolige — APPROVED** [review]

**Review by csullivan — APPROVED** [review]

**Comment by masahi** [comment]
Great `set_source_files_properties` worked! I removed `rocthrust/thrust.cc`, the same `thrust.cu` is used by CUDA and rocm now cc @mbrookhart

**Review by mbrookhart — APPROVED** [review]

**Comment by masahi** [comment]
Thanks @mbrookhart @tkonolige @csullivan

**Pull request merged**